Sources/NIO/ByteBuffer-extract.swift

```diff
@@ -0,0 +1,197 @@
+    
+    let byteBuffer: ByteBuffer
+    let separator: UInt8
+    let start: Int
+    let length: Int
+    
+    private var currentIndex: Int
+    
+    public typealias Element = ByteBuffer
+    mutating public func next() -> ByteBuffer? {
+        
+        let slicingParameters = byteBuffer.withVeryUnsafeBytes { pointer -> (start: Int, length: Int) in
+            
+            let startPoint = pointer.baseAddress!.assumingMemoryBound(to: UInt8.self)
+                                    .advanced(by: self.start)
+            
+            var address = startPoint.advanced(by: currentIndex)
+            let finalAddress = startPoint.advanced(by: length)
+            
+            let initialPointer = address
+
+            while address < finalAddress {
+                
+                if address.pointee == separator {
+                    break
+                }
+                address = address.advanced(by: 1)
+            }
+            
+            let start = currentIndex
+            let tokenLength = address - initialPointer
+            // For the next time, with skipping the separator
+            currentIndex += tokenLength + 1
+            
+            return (start: start, length: tokenLength)
+        }
+        return byteBuffer.getSlice(at: slicingParameters.start + self.start,
+                                   length: slicingParameters.length)
+
+    }
+    
+    public init(byteBuffer: ByteBuffer, separator: UInt8, start: Int, length: Int) {
+        precondition(start <= byteBuffer.capacity - length)
+        
+        self.byteBuffer = byteBuffer
+        self.separator = separator
+        self.start = start
+        self.length = length
+        self.currentIndex = 0
+    }
+
+    public init(byteBuffer: ByteBuffer,
+                separator: UInt8) {
+        
+        self.init(byteBuffer: byteBuffer, separator: separator,
+                  start: byteBuffer.readerIndex, length: byteBuffer.readableBytes)
+    }
+}
+
+extension ByteBuffer {
+    
+    static var defaultWhitespaces = [" ", "\t"].map({UInt8($0.utf8CString[0])})
+    
+    public func sliceByTrimming(whiteSpaces: [UInt8], start: Int, length: Int) -> ByteBuffer? {
+        return withVeryUnsafeBytes { pointer -> ByteBuffer? in
+            assert(start <= self.capacity - length)
+            
+            let firstPtr = pointer.baseAddress!.assumingMemoryBound(to: UInt8.self)
+
+            var startPtr = firstPtr.advanced(by: start)
+            var endPtr   = startPtr.advanced(by: length - 1)
+            
+            // Advance startPtr until not a whiteSpace
+            while startPtr < endPtr && startPtr.pointee != 0 && whiteSpaces.contains(startPtr.pointee) {
+                startPtr = startPtr.advanced(by: 1)
+            }
+            
+            // Retreat endPtr until not a whiteSpace, null is considered a whiteSpace
+            while endPtr >= startPtr && (whiteSpaces.contains(endPtr.pointee) || endPtr.pointee == 0) {
```

Comment:
These loops can also take advantage of buffer pointers to avoid manual offset management.